crates/bevy_reflect/src/type_registry.rs

```diff
@@ -100,39 +107,114 @@ impl TypeRegistry {
         registry
     }
 
-    /// Registers the type `T`, adding reflect data as specified in the [`Reflect`] derive:
-    /// ```ignore (Neither bevy_ecs nor serde "derive" are available.)
-    /// #[derive(Component, serde::Serialize, serde::Deserialize, Reflect)]
-    /// #[reflect(Component, Serialize, Deserialize)] // will register ReflectComponent, ReflectSerialize, ReflectDeserialize
+    /// Attempts to register the type `T` if it has not yet been registered already.
+    ///
+    /// This will also recursively register any type dependencies as specified by [`GetTypeRegistration::register_type_dependencies`].
+    /// When deriving `Reflect`, this will generally be all the fields of the struct or enum variant.
+    /// As with any type registration, these type dependencies will not be registered more than once.
+    ///
+    /// If the registration for type `T` already exists, it will not be registered again and neither will its type dependencies.
+    /// To register the type, overwriting any existing registration, use [register](Self::overwrite_registration) instead.
+    ///
+    /// Additionally, this will add any reflect [type data](TypeData) as specified in the [`Reflect`] derive.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use std::any::TypeId;
+    /// # use bevy_reflect::{Reflect, TypeRegistry, std_traits::ReflectDefault};
+    /// #[derive(Reflect, Default)]
+    /// #[reflect(Default)]
+    /// struct Foo {
+    ///   name: Option<String>,
+    ///   value: i32
+    /// }
+    ///
+    /// let mut type_registry = TypeRegistry::default();
+    ///
+    /// type_registry.register::<Foo>();
+    ///
+    /// // The main type
+    /// assert!(type_registry.contains(TypeId::of::<Foo>()));
+    ///
+    /// // Its type dependencies
+    /// assert!(type_registry.contains(TypeId::of::<Option<String>>()));
+    /// assert!(type_registry.contains(TypeId::of::<i32>()));
+    ///
+    /// // Its type data
+    /// assert!(type_registry.get_type_data::<ReflectDefault>(TypeId::of::<Foo>()).is_some());
     /// ```
     pub fn register<T>(&mut self)
     where
         T: GetTypeRegistration,
     {
-        self.add_registration(T::get_type_registration());
+        if self.add_registration(T::get_type_registration()) {
+            T::register_type_dependencies(self);
+        }
```

Comment:
@cart Would this possibly incur more `TypeId` hashes than the [previous version](https://github.com/bevyengine/bevy/blob/799a23a93a000b4fb764f0984cff3f75b2544b3d/crates/bevy_reflect/src/type_registry.rs#L151-L156)?

This will call `T::get_type_registration()` every time. And most `GetTypeRegistration` impls will have at least 1 `TypeId` hash via `TypeRegistration::insert` (`ReflectFromPtr`) but more often at least 2 (`ReflectFromReflect`). 

That's the minimum, but most types are registering other type data, such as `ReflectComponent`, `ReflectDefault`, `ReflectSerialize`, etc.

So each call to `register` now will typically require a few extra hashes rather than just two.

Reply:
Ew yeah wasn't thinking about the cost of `get_type_registration()` at all.

I think this would work:

1. Rework add_registration to accept TypeId and a `fn() -> TypeRegistration`. Call the `fn` only when actually needed internally
2. Call TypeId::of::<T>() directly in `register()`

Reply:
I agree, eagerly computing `T::get_type_registration` is likely not good for performance. Maybe it would be better to add a variant of `add_registration` that takes a closure and only calls it if the registration is not already present?

Reply:
I'll try it out. I think it would make sense to make this a private function so we don't allow `add_registration` to be abused (i.e. registering types that are not allowed to be registered).

Reply:
Okay I [added `register_internal`](https://github.com/bevyengine/bevy/pull/5781/commits/e9faa7961ddfa4b207534b6e54159a0bbc6befa1) to do just that!